File: src/python/gpmds/GPMDS.py
#!/usr/bin/env python
from numpy import *
import numpy as np
import math

from gpmds.MultiGPR import MultiGPR


class GPMDS:
    def __init__(self, ell, sigmaF, sigmaN, velocity_cap=None):
        self.ell = ell
        self.sigmaF = sigmaF
        self.sigmaN = sigmaN

        self.mGPR = MultiGPR(2,2)

        self.set_gp_parameters(ell, sigmaF, sigmaN)
        if not velocity_cap:
            velocity_cap = (0.1, 1.0)

        self.v_capLow = velocity_cap[0]
        self.v_capHigh = velocity_cap[1]



    def set_gp_parameters(self, ell, sigmaF, sigmaN):
        self.mGPR.setHyperParams(ell, sigmaF, sigmaN)

    def setOriginalDynamics(self, fun):
        # Set the original dynamics to the user passed function
        self.originalDynamics = fun

    def addData(self, position, velocity):
        theta = self.getTheta(position, velocity)
        # Add the data point to training data
        self.mGPR.addTrainingData(position, theta)

    def addDataM(self, position, velocity):
        for i in range(0,size(position[0])):
            self.addData(array([position[0,i],position[1,i]]), array([velocity[0,i],velocity[1,i]]))

        self.mGPR.prepareRegression()


    def getTheta(self, position, velocity):
        # Compute original dynamics at position
        originalVelocity = self.originalDynamics(position)
        # Find scaling between vectors
        kappa = (linalg.norm(velocity)/linalg.norm(originalVelocity))-1.0

        # Compare original velocity with velocity to compute speed scaling (kappa) and rotation axis angle
        # Check for divide by 0 error
        if linalg.norm(velocity)==0:
            normDesDir = velocity
        else:
            normDesDir = velocity/linalg.norm(velocity)

        #print originalVelocity
        if linalg.norm(originalVelocity)==0:
            normOrgDir = originalVelocity
        else:
            normOrgDir = originalVelocity/linalg.norm(originalVelocity)

        # Calculate angle of rotation
        angle = math.atan2(normDesDir[1],normDesDir[0])-math.atan2(normOrgDir[1],normOrgDir[0])

        # put in a good range
        if(angle > pi):
            angle = -(2*pi-angle)
        elif(angle < -pi):
            angle = 2*pi+angle;

        # Create the datapoint, theta (the 2 number output of the GPR- angle, scaling)
        theta = zeros(2)
        theta[0] = angle
        theta[1] = kappa

        return theta


    def removeData(self, start, end):#position, velocity):
        #print 'Removing data...'
        #theta = self.getTheta(position, velocity)
        s =  self.mGPR.removeTrainingData(start, end)#self.mGPR.removeTrainingData(position,theta)
        if len(self.mGPR.inputData) > 0:
            self.mGPR.prepareRegression()
        return s

    def reshapedDynamics(self, position):
        # Get original dynamics at position
        originalVelocity = self.originalDynamics(position)
        if self.mGPR.nData==0:
            velocity = originalVelocity
            #if(linalg.norm(velocity)>self.v_capHigh):
            #    velocity = velocity/linalg.norm(velocity)*self.v_capHigh
            #if (linalg.norm(velocity)<self.v_capLow):
            #    velocity = velocity/linalg.norm(velocity)*self.v_capLow
            #return velocity
        else:
            result = self.mGPR.doRegression(position)
            angle = result[0]
            kappa = result[1]
            kappa = max(kappa, -0.9)

            # Calculate the rotation matrix
            R = array([[cos(angle), -sin(angle)],[sin(angle),cos(angle)]])
            # Rotate velocity by R and scale by kappa
            velocity = (1.0 + kappa)*dot(R, originalVelocity)

        # Put velocity limits
        if (linalg.norm(velocity)>self.v_capHigh):
            velocity = velocity/linalg.norm(velocity)*self.v_capHigh
        if (linalg.norm(velocity)<self.v_capLow):
            velocity = velocity/linalg.norm(velocity)*self.v_capLow
        return velocity



def originalDynamicsLinear(pose):
    p = 4
    A = -p * np.eye(2)
    xd = np.dot(A, pose)
    jacobian = A
    return xd
# function [xd jacobian] = originalDynamicsLINEAR(x)
# % stable linear isotropic dynamics. adjust speed of convergence with p.
# p = 4;
# A = -p*eye(2);
# xd =A*x;
# %xd(2,:) =xd(2,:) + 2*p*x(1,:).*cos(2*pi/150*x(1,:));
# jacobian = A;


if __name__ == '__main__':
    print 'hello'

    (ell, sigmaF, sigmaN) = (30, 1.0, 0.4)
    gp = GPMDS(ell, sigmaF, sigmaN)
    print gp
    print gp.mGPR

    originalDynamics = originalDynamicsLinear

    pose = np.array((5, 2))
    velocity = np.array((1.0, 1.0))

    gp.setOriginalDynamics(originalDynamics)

    gp.addData(pose, velocity)
    gp.mGPR.prepareRegression()

    th = gp.getTheta(pose, velocity)
    print 'Th: {}'.format(th)

    reshaped = gp.reshapedDynamics(pose)
    print 'Reshaped: {}'.format(reshaped)

File: src/cpp/GPMDS/src/GPMDS.cpp
#include "GPMDS.h"
#include <math.h>
#include "Timer.h"

GPMDS::GPMDS(REALTYPE ell, REALTYPE sigmaF, REALTYPE sigmaN,REALTYPE speedErrorTol, REALTYPE angleErrorTol)
{
  //Initialize Gaussian Process
  setGPParameters(ell, sigmaF, sigmaN);
  this->speedErrorTol = speedErrorTol;
  this->angleErrorTol = angleErrorTol;
}

void GPMDS::setGPParameters(REALTYPE ell, REALTYPE sigmaF, REALTYPE sigmaN)
{
  //Create the GPR (input 3, output 4)
  mGPR = new MultiGPR(3,4);
  //Set parameters for the GPR;
  (*mGPR).setHyperParams(ell, sigmaF, sigmaN);
}

void GPMDS::setOriginalDynamics(Vector3r (*fun)(Vector3r))
{
  //Set the original dynamics to the function passed in by the user
  originalDynamics = fun;
}

void GPMDS::prepareFastQuery()
{
  //Set the original dynamics to the function passed in by the user
  mGPR->prepareRegression(true);
}

Vector4r GPMDS::computeLMDSParameters(Vector3r position, Vector3r velocity){
  Vector3r originalVelocity, axis;
  Vector4r theta;
  REALTYPE kappa, angle;

  //Compute original dynamics at position
  originalVelocity = originalDynamics(position);

  // Find the scaling between vectors
  kappa = (velocity.norm()/originalVelocity.norm())-1.0;

  //Compare resulting original velocity with velocity to compute speed scaling (kappa) and rotation axis angle
  angle = acos((originalVelocity.dot(velocity))/(originalVelocity.norm()*velocity.norm()));
  axis = originalVelocity.cross(velocity);
  axis /= axis.norm();

  //Create the datapoint, theta (the 4 number output of the GPR- axis angle, scaling)
  for (int i=0; i<3; i++)
    theta(i) = angle*axis(i);

  theta(3) = kappa;

  return theta;
}


void GPMDS::addData(Vector3r position, Vector3r velocity)
{
  Vector4r theta;
  theta = computeLMDSParameters(position, velocity);
  if(checkNewData(position,theta))
    mGPR->addTrainingData(position, theta);
}

bool GPMDS::checkNewData(Vector3r position, Vector4r theta){
  if(mGPR->getNData() <1)
    return true;

  Vector4r theta_p;
  theta_p = mGPR->doRegression(position,true);
  Vector3r angleaxis_p,angleaxis;
  for(int i=0;i<3;i++){
    angleaxis(i) = theta(i);
    angleaxis_p(i) = theta_p(i);
  }
  REALTYPE angle,angle_p;
  angle_p = angleaxis_p.norm();
  angle = angleaxis.norm();
  if(fabs(theta_p(3) - theta(3))>speedErrorTol)// || fabs(angle_p-angle)>angleErrorTol){
  {
    return true;
  }
  else{
    std::cout << mGPR->getNData() << std::endl;
    return false;
  }
}

Vector3r GPMDS::reshapedDynamics(Vector3r position)
{

  Vector3r originalVelocity, velocity, axis;
  Vector4r result;
  Matrix3r rot_mat;
  REALTYPE kappa, angle;


  // Get original dynamics at position
  originalVelocity = originalDynamics(position);

  // Perform GPR on the given position to compute reshaping parameters
  //(*mGPR).prepareRegression();
  //cout<<originalVelocity<<endl;
  result = (*mGPR).doRegression(position);

  // Calculate axis and angle from reshaping parameters
  for (int i=0; i<3; i++){axis(i)=result(i);}
  kappa  = result(3);
  angle = axis.norm();
  axis = axis/angle;
  if (angle < 0.001)
  {
    // If the angle is very small, make rotation matrix the identity
    rot_mat.Identity();
  }
  else
  {
    // Find the rotation matrix from the axis and angle
    AngleAxisr rot_aa(angle, axis);
    rot_mat = rot_aa.toRotationMatrix();
  }
  //Rotate and scale the original velocity
  velocity = rot_mat*originalVelocity;
  velocity= velocity*(kappa+1);

  //Return resulting velocity
  return velocity;
}
